frontend/model/notifications/nativeNotification.js

```diff
@@ -4,50 +4,115 @@ import sbp from '@sbp/sbp'
+    typeof ServiceWorker !== 'function' ||
+    typeof navigator.serviceWorker !== 'object'
+  ) return
 
-  if (Notification.permission === 'granted') {
-    await sbp('service-worker/setup-push-subscription').catch((e) => {
-      console.error('Error setting up service worker', e)
+  if (
+    typeof navigator.permissions === 'object' &&
+    // $FlowFixMe[method-unbinding]
+    typeof navigator.permissions.query === 'function'
+  ) {
+    Promise.all([
+      navigator.permissions.query({ name: 'notifications' }),
+      navigator.permissions.query({ name: 'push', userVisibleOnly: true })
+    ]).then(
+      (statuses) => {
+        const states = statuses.map(status => status.state)
+        handler(states)
+        statuses[0].addEventListener('change', () => {
+          states[0] = statuses[0].state
+          handler(states)
+        }, false)
+        statuses[1].addEventListener('change', () => {
+          states[1] = statuses[1].state
+          handler(states)
+        }, false)
+      }
+    ).catch((e) => {
+      console.error('Error querying notifications permission', e)
+      fallbackChangeListener()
     })
+  } else {
+    fallbackChangeListener()
+  }
+}
+
+export async function requestNotificationPermission (): Promise<null | string> {
+  if (typeof Notification !== 'function') {
+    return null
   }
 
-  return Notification.permission
+  try {
+    const result = await Notification.requestPermission()
+    sbp('state/vuex/commit', 'setNotificationEnabled', result === 'granted')
+    return result
+  } catch (e) {
+    console.error('requestNotificationPermission:', e.message)
+    return null
+  }
 }
 
 export function makeNotification ({ title, body, icon, path }: {
   title: string, body: string, icon?: string, path?: string
 }): void {
-  if (Notification?.permission === 'granted' && sbp('state/vuex/settings').notificationEnabled) {
+  if (typeof Notification !== 'function') return
+  // If not running on a SW
+  if (typeof WorkerGlobalScope !== 'function') {
     try {
+      // $FlowFixMe[incompatible-type]
+      if (navigator.vendor === 'Apple Computer, Inc.') {
+        throw new Error('Safari requires a service worker for the notification to be displayed')
+      }
       const notification = new Notification(title, { body, icon })
       if (path) {
         notification.onclick = (event) => {
           sbp('controller/router').push({ path }).catch(console.warn)
         }
       }
-    } catch {
-      try {
-        // FIXME: find a cross-browser way to pass the 'path' parameter when the notification is clicked.
-        navigator.serviceWorker?.ready.then(registration => {
-          // $FlowFixMe
-          registration.showNotification(title, { body, icon })
-        }).catch(console.warn)
-      } catch (error) {
-        console.error('makeNotification: ', error.message)
-      }
+    } catch (e) {
+      return navigator.serviceWorker?.ready.then(registration => {
```

Comment:
```
Error -------------------------------------------------------- frontend/model/notifications/nativeNotification.js:103:14

Cannot return `navigator.serviceWorker?.ready.then(...).catch(...)` because `Promise` [1] is incompatible with
undefined [2]. [incompatible-return]

   frontend/model/notifications/nativeNotification.js:103:14
                      v----------------------------------------------------
    103|       return navigator.serviceWorker?.ready.then(registration => {
    104|         // $FlowFixMe
    105|         return registration.showNotification(title, { body, icon, data: { path } })
    106|       }).catch(console.warn)
               ---------------------^
```

Reply:
Fixed